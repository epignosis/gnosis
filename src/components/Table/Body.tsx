import React, { FC, useState } from "react";
import Checkbox from "../FormElements/CheckboxGroup/Checkbox";
import Result from "../Result/Result";
import Cell from "./Cell";
import { ChildrenProps } from "./Table";
import { Row } from "types/types";

const Body: FC<ChildrenProps> = ({ state, dispatch }) => {
  const [hoveredRow, setHoveredRow] = useState("");
  const { selectable, columns, selected, handleRowClick } = state;
  const accessors = columns
    .filter((column) => !column.hidden)
    .map((column) => column.accessor)
    .filter((column) => column !== "actions");
  const selectedRows = selected.map((entry) => entry.id);

  const handleRowSelection = (e: React.ChangeEvent<HTMLInputElement>, row: Row): void => {
    e.preventDefault();

    if (!selectedRows.includes(row.id)) {
      dispatch({ type: "SELECT_ROW", payload: row });
    }

    if (selectedRows.includes(row.id)) {
      dispatch({ type: "REMOVE_ROW", payload: row });
    }
  };

  return (
    <tbody>
      {state.rows.length > 0 ? (
        <>
          {state.rows.map((row) => {
            const isSelected = selectedRows.includes(row.id);
            const hasClickHandler = handleRowClick ? "link" : "";
            const isSelectedClassname = isSelected ? "selected" : "";
            const isHoveredRow = row.actions && hoveredRow === `${row.id}`;

            return (
              <tr
                key={row.id}
                className={`${isSelectedClassname} ${hasClickHandler}`}
                onMouseEnter={(): void => setHoveredRow(row.id.toString())}
                onMouseLeave={(): void => setHoveredRow("")}
              >
                {selectable && (
                  <Cell key={row.id} className="selectable-cell">
                    <Checkbox
                      id={`entry-${row.id}`}
                      name={`entry-${row.id}`}
                      value={`entry-${row.id}`}
                      onChange={(e): void => handleRowSelection(e, row)}
                      checked={isSelected}
                    />
                  </Cell>
                )}

                {accessors.map((accessor) => {
                  const rowObj = row[accessor];
                  const hasClickHandler = handleRowClick
                    ? (): void => handleRowClick(row)
                    : undefined;

                  if (typeof rowObj === "function") {
                    return (
                      <Cell key={`entry-${row.id}-${accessor}`} onClick={hasClickHandler}>
                        {rowObj(row)}
                      </Cell>
                    );
                  }
                  return (
                    <Cell key={`entry-${row.id}-${accessor}`} onClick={hasClickHandler}>
                      {rowObj as string}
                    </Cell>
                  );
                })}

                {isHoveredRow && (
                  <td className="actions-container">
                    {typeof row?.actions === "function" && row?.actions(row)}
                  </td>
                )}
              </tr>
            );
          })}
        </>
      ) : (
        <tr className="empty-state-container">
          <td>
            <Result title={state.emptyState.title} info={state.emptyState.info} />
          </td>
        </tr>
      )}
    </tbody>
  );
};

export default Body;
