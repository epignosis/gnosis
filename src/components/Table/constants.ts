export enum Actions {
  toggleAll = "TOGGLE_ALL",
  toggle = "TOGGLE",
  sortingChanged = "SORTING_CHANGED",
  columnsChanged = "COLUMNS_CHANGED",
  rowsChanged = "ROWS_CHANGED",
  resetSelectedRows = "RESET",
  setDisabled = "SET_DISABLED",
  selectMany = "SELECT_MANY",
}
