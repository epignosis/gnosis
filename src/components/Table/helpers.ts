import { Column } from "./types";

export const getVisibleAccessors = (columns: Column[]): string[] => {
  return columns.filter((column) => !column.hidden).map((column) => column.accessor);
};

export const getDefaultAccessor = (columns: Column[]): string | undefined => {
  const visibleAccessors = getVisibleAccessors(columns);
  return columns.find((column) => column.isDefaultAccessor)?.accessor ?? visibleAccessors[0];
};
