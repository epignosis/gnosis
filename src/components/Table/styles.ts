import { css } from "@emotion/react";

export const tableContainer = css`
  border: 1px solid #dddddd;
  border-radius: 5px;
  overflow: hidden;

  .table-container {
    overflow-x: auto;

    table {
      width: 100%;
      background-color: #ffffff;
      border-spacing: 0;
      white-space: nowrap;

      thead {
        box-shadow: 0px 1px 14px #dddddd;

        tr {
          height: 5rem;

          th {
            font-weight: 700;
            padding: 0 1.5rem;
            text-align: left;
          }
        }
      }

      tbody {
        tr {
          &:first-of-type {
            td {
              padding-top: 1.5rem;
            }
          }

          &:last-of-type {
            td {
              border: 0;
            }
          }
        }

        td {
          padding: 1rem 1.5rem;
          border-bottom: 1px solid #dddddd;
        }
      }
    }
  }

  footer {
    box-shadow: 0px 2px 14px #dddddd;
    padding: 0.75rem 1.5rem;
    text-align: center;
  }
`;
