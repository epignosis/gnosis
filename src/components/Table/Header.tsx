import React, { FC } from "react";
import Checkbox from "../FormElements/CheckboxGroup/Checkbox";
import { IconChevronDownSVG, IconChevronUpSVG } from "../../icons/index";
import Cell from "./Cell";
import { ChildrenProps } from "./Table";
import { Sorting } from "types/types";

const Header: FC<ChildrenProps> = ({ state, dispatch }) => {
  const { rows, selectable, columns, selected, sortable, sorting } = state;
  const selectedIds = selected.map((entry) => entry.id);
  const rowIds = rows.map((row) => row.id);
  const isSelectAllChecked = selected.length > 0;
  const allRowsSelected = rowIds.every((rowId) => selectedIds.includes(rowId));

  const handleCheckboxClick = (e: React.ChangeEvent<HTMLInputElement>): void => {
    e.preventDefault();
    selected.length === 0 ? dispatch({ type: "SELECT_ALL" }) : dispatch({ type: "REMOVE_ALL" });
  };

  const handleSortingChange = (accesor: string): void => {
    if (sortable) {
      // new sorting object
      const newSorting: Sorting = {
        column: accesor,
        isDescending: false,
      };

      // sorting the same column
      if (sorting?.column === accesor) {
        !sorting.isDescending ? (newSorting.isDescending = true) : (newSorting.column = "");
      }

      dispatch({ type: "SORTING_CHANGED", payload: newSorting });
    }
  };

  return (
    <thead>
      <tr>
        {selectable && (
          <Cell as="th" key={`select-all-${isSelectAllChecked}`} className="selectable-cell">
            <Checkbox
              id="select-all"
              name="select-all"
              value="all"
              onChange={handleCheckboxClick}
              checked={isSelectAllChecked}
              isPartiallySelected={!allRowsSelected}
            />
          </Cell>
        )}
        {columns.map(
          ({ accessor, cell, hidden, sortableHeader = true, classNames = [] }) =>
            !hidden && (
              <Cell
                as="th"
                key={accessor}
                className={`header-cell ${classNames.length > 0 && classNames.join(" ")}`}
                onClick={(): void => {
                  if (sortableHeader) {
                    handleSortingChange(accessor);
                  }
                }}
              >
                <span>{typeof cell === "string" ? cell : cell({ accessor, cell })}</span>
                {sortable && sorting?.column === accessor && (
                  <span className="sorting-icon">
                    {!sorting?.isDescending ? (
                      <IconChevronUpSVG height={20} />
                    ) : (
                      <IconChevronDownSVG height={20} />
                    )}
                  </span>
                )}
              </Cell>
            ),
        )}
      </tr>
    </thead>
  );
};

export default Header;
