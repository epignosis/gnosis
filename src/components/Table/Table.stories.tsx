import React from "react";
import { Story } from "@storybook/react";
import { WithBothBtns } from "../Pagination/Pagination.stories";
import { PaginationProps } from "../Pagination/Pagination";
import Table, { TableProps } from "./Table";

export default {
  title: "components/Table",
  component: Table,
};

const Template: Story<TableProps> = (args) => <Table {...args} />;

export const Default = Template.bind({});

Default.args = {
  columns: [
    {
      Header: "Language",
      accessor: "language",
      // eslint-disable-next-line react/display-name
      Cell: ({ value }): JSX.Element => <span>~ {value} ~</span>,
    },
    {
      Header: "Love level",
      accessor: "love_level",
    },
    {
      Header: "Awesomeness",
      accessor: "awesomeness",
    },
    {
      Header: "Favorite framework",
      accessor: "favorite_framework",
    },
    {
      accessor: "action",
      disableSortBy: true,
    },
  ],
  data: [
    {
      language: "Rust",
      love_level: "To infinity and beyond!",
      awesomeness: "Super awesome",
      favorite_framework: (
        <a href="https://github.com/tokio-rs/tokio" target="_blank" rel="noreferrer">
          tokio-rs/tokio
        </a>
      ),
      action: <button>&hearts;&hearts;&hearts;</button>,
    },
    {
      language: "Typescript",
      love_level: "Much love",
      awesomeness: "Such maze",
      favorite_framework: (
        <a href="https://reactjs.org/" target="_blank" rel="noreferrer">
          React
        </a>
      ),
      action: <button>&hearts;&hearts;&hearts;</button>,
    },
    {
      language: "PHP",
      love_level: "You'll find no such thing here",
      awesomeness: "Dino awesome",
      favorite_framework: "None",
      action: "No action...",
    },
    {
      language: "Rust",
      love_level: "To infinity and beyond!",
      awesomeness: "Super awesome",
      favorite_framework: (
        <a href="https://github.com/tokio-rs/tokio" target="_blank" rel="noreferrer">
          tokio-rs/tokio
        </a>
      ),
      action: <button>&hearts;&hearts;&hearts;</button>,
    },
    {
      language: "Typescript",
      love_level: "Much love",
      awesomeness: "Such maze",
      favorite_framework: (
        <a href="https://reactjs.org/" target="_blank" rel="noreferrer">
          React
        </a>
      ),
      action: <button>&hearts;&hearts;&hearts;</button>,
    },
    {
      language: "PHP",
      love_level: "You'll find no such thing here",
      awesomeness: "Dino awesome",
      favorite_framework: "None",
      action: "No action...",
    },
  ],
};

Default.argTypes = {
  footer: {
    control: false,
  },
};

export const WithFooter = Template.bind({});

WithFooter.args = {
  ...Default.args,
  footer: <WithBothBtns {...(WithBothBtns.args as PaginationProps)} />,
};
