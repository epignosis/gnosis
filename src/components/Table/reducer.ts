import { Actions } from "./constants";
import { ActionType, TableState } from "./types";

export type Dispatch = (action: ActionType) => void;

export const reducer = (state: TableState, action: ActionType): TableState => {
  switch (action.type) {
    case Actions.selectAll: {
      return { ...state, selected: [...state.rows] };
    }
    case Actions.removeAll: {
      return { ...state, selected: [] };
    }
    case Actions.selectRow: {
      return { ...state, selected: [...(state.selected ?? []), action.payload] };
    }
    case Actions.removeRow: {
      const rowToRemoveId = action.payload.id;
      const newSelectedRow = state.selected?.filter((row) => row?.id !== rowToRemoveId);

      return { ...state, selected: newSelectedRow };
    }
    case Actions.sortingChanged: {
      if (state.sorting && state?.onSortingChanged) {
        state.onSortingChanged(action.payload);
      }

      return { ...state, sorting: action.payload };
    }
    case Actions.columnsChanged: {
      return { ...state, columns: action.payload };
    }
    case Actions.rowsChanged: {
      return { ...state, rows: action.payload };
    }
    default: {
      return state;
    }
  }
};
