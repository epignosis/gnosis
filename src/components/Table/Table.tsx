/* eslint-disable react/jsx-key */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { FC, ReactNode, useMemo } from "react";
import { useTable, useSortBy, Column } from "react-table";
import { tableContainer } from "./styles";

export type TableProps = {
  columns: (Column & { disableSortBy?: boolean })[];
  data: unknown[];
  footer?: ReactNode;
};

const Table: FC<TableProps> = ({ columns, data, footer }) => {
  const memoizedColumns = useMemo(() => columns, [columns]);
  const memoizedData = useMemo(() => data, [data]);
  const { getTableProps, getTableBodyProps, headerGroups, rows, prepareRow } = useTable(
    {
      columns: memoizedColumns,
      data: memoizedData as any,
    },
    useSortBy,
  );

  return (
    <div css={tableContainer}>
      <div className="table-container">
        <table {...getTableProps()}>
          <thead>
            {
              // Loop over the header rows
              headerGroups.map((headerGroup) => (
                // Apply the header row props
                <tr {...headerGroup.getHeaderGroupProps()}>
                  {
                    // Loop over the headers in each row
                    headerGroup.headers.map((column) => (
                      // Apply the header cell props
                      <th {...column.getHeaderProps()}>{column.render("Header")}</th>
                    ))
                  }
                </tr>
              ))
            }
          </thead>
          {/* Apply the table body props */}
          <tbody {...getTableBodyProps()}>
            {
              // Loop over the table rows
              rows.map((row) => {
                // Prepare the row for display
                prepareRow(row);
                return (
                  // Apply the row props
                  <tr {...row.getRowProps()} data-testid="body-row">
                    {
                      // Loop over the rows cells
                      row.cells.map((cell) => {
                        // Apply the cell props
                        return (
                          <td {...cell.getCellProps()}>
                            {
                              // Render the cell contents
                              cell.render("Cell")
                            }
                          </td>
                        );
                      })
                    }
                  </tr>
                );
              })
            }
          </tbody>
        </table>
      </div>
      {footer && <footer>{footer}</footer>}
    </div>
  );
};

export default Table;
