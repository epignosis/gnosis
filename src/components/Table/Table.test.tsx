import React from "react";
import Table from "./Table";
import { screen, render } from "@test-utils/render";

const columns = [
  {
    Header: "Love level",
    accessor: "love_level",
  },
  {
    Header: "Awesomeness",
    accessor: "awesomeness",
  },
  {
    Header: "Favorite framework",
    accessor: "favorite_framework",
  },
];

const data = [
  {
    love_level: "You'll find no such thing here",
    awesomeness: "Dino awesome",
    favorite_framework: "None",
  },
  {
    love_level: "To infinity and beyond!",
    awesomeness: "Super awesome",
    favorite_framework: (
      <a href="https://github.com/tokio-rs/tokio" target="_blank" rel="noreferrer">
        tokio-rs/tokio
      </a>
    ),
  },
  {
    love_level: "Much love",
    awesomeness: "Such maze",
    favorite_framework: (
      <a href="https://reactjs.org/" target="_blank" rel="noreferrer">
        React
      </a>
    ),
  },
  {
    love_level: "You'll find no such thing here",
    awesomeness: "Dino awesome",
    favorite_framework: "None",
  },
];

describe("<Table>", () => {
  it("renders correctly without Footer", () => {
    render(<Table columns={columns} data={data} />);

    const footer = screen.queryByRole("footer");
    const headerTitles = screen.getAllByRole("columnheader");
    const bodyRows = screen.getAllByTestId("body-row");
    const cells = screen.getAllByRole("cell");

    expect(footer).not.toBeInTheDocument();
    expect(headerTitles).toHaveLength(3);
    expect(bodyRows).toHaveLength(4);
    expect(cells).toHaveLength(headerTitles.length * bodyRows.length);
  });

  it("renders correctly with Footer", () => {
    render(<Table columns={columns} data={data} footer={<div>Test footer</div>} />);

    const footer = screen.getByText("Test footer");
    expect(footer).toHaveTextContent("Test footer");
  });

  it("matches snapshot without footer", () => {
    const { container } = render(<Table columns={columns} data={data} />);

    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with footer", () => {
    const { container } = render(
      <Table columns={columns} data={data} footer={<div>Test footer</div>} />,
    );

    expect(container).toMatchSnapshot();
  });
});
