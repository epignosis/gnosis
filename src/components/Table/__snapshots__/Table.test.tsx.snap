// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table> matches snapshot with footer 1`] = `
<div>
  <div
    class="css-1ap5vgl-tableContainer"
  >
    <div
      class="table-container"
    >
      <table
        role="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              colspan="1"
              role="columnheader"
            >
              Love level
            </th>
            <th
              colspan="1"
              role="columnheader"
            >
              Awesomeness
            </th>
            <th
              colspan="1"
              role="columnheader"
            >
              Favorite framework
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              You'll find no such thing here
            </td>
            <td
              role="cell"
            >
              Dino awesome
            </td>
            <td
              role="cell"
            >
              None
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              To infinity and beyond!
            </td>
            <td
              role="cell"
            >
              Super awesome
            </td>
            <td
              role="cell"
            >
              <a
                href="https://github.com/tokio-rs/tokio"
                rel="noreferrer"
                target="_blank"
              >
                tokio-rs/tokio
              </a>
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              Much love
            </td>
            <td
              role="cell"
            >
              Such maze
            </td>
            <td
              role="cell"
            >
              <a
                href="https://reactjs.org/"
                rel="noreferrer"
                target="_blank"
              >
                React
              </a>
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              You'll find no such thing here
            </td>
            <td
              role="cell"
            >
              Dino awesome
            </td>
            <td
              role="cell"
            >
              None
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <footer>
      <div>
        Test footer
      </div>
    </footer>
  </div>
</div>
`;

exports[`<Table> matches snapshot without footer 1`] = `
<div>
  <div
    class="css-1ap5vgl-tableContainer"
  >
    <div
      class="table-container"
    >
      <table
        role="table"
      >
        <thead>
          <tr
            role="row"
          >
            <th
              colspan="1"
              role="columnheader"
            >
              Love level
            </th>
            <th
              colspan="1"
              role="columnheader"
            >
              Awesomeness
            </th>
            <th
              colspan="1"
              role="columnheader"
            >
              Favorite framework
            </th>
          </tr>
        </thead>
        <tbody
          role="rowgroup"
        >
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              You'll find no such thing here
            </td>
            <td
              role="cell"
            >
              Dino awesome
            </td>
            <td
              role="cell"
            >
              None
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              To infinity and beyond!
            </td>
            <td
              role="cell"
            >
              Super awesome
            </td>
            <td
              role="cell"
            >
              <a
                href="https://github.com/tokio-rs/tokio"
                rel="noreferrer"
                target="_blank"
              >
                tokio-rs/tokio
              </a>
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              Much love
            </td>
            <td
              role="cell"
            >
              Such maze
            </td>
            <td
              role="cell"
            >
              <a
                href="https://reactjs.org/"
                rel="noreferrer"
                target="_blank"
              >
                React
              </a>
            </td>
          </tr>
          <tr
            data-testid="body-row"
            role="row"
          >
            <td
              role="cell"
            >
              You'll find no such thing here
            </td>
            <td
              role="cell"
            >
              Dino awesome
            </td>
            <td
              role="cell"
            >
              None
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
